src/Symfony/Component/Lock/Store/MysqlStore.php

```diff
@@ -0,0 +1,141 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\InvalidArgumentException;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\PersistingStoreInterface;
+
+/**
+ * @author rtek
+ * @author Jérôme TAMARELLE <jerome@tamarelle.net>
+ */
+class MysqlStore implements PersistingStoreInterface
+{
+    private \PDO $conn;
+
+    private string $dsn;
+
+    private array $options;
+
+    private int $connectionId;
+
+    /** @var bool[] */
+    private static array $locksAcquired = [];
+
+    public function __construct(\PDO|string $connOrDsn, array $options = [])
+    {
+        if ($connOrDsn instanceof \PDO) {
+            $this->conn = $connOrDsn;
+            $this->assertMysqlDriver();
+            if (\PDO::ERRMODE_EXCEPTION !== $this->conn->getAttribute(\PDO::ATTR_ERRMODE)) {
+                throw new InvalidArgumentException(sprintf('"%s" requires PDO error mode attribute be set to throw Exceptions (i.e. $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION)).', __METHOD__));
+            }
+        } else {
+            $this->dsn = $connOrDsn;
+        }
+
+        $this->options = $options;
+    }
+
+    public function save(Key $key): void
+    {
+        $id = $this->getLockId($key);
+
+        if (self::$locksAcquired[$id] ?? false) {
+            return;
+        }
+
+        $stmt = $this->conn->prepare('SELECT IF(IS_USED_LOCK(:name) = CONNECTION_ID(), -1, GET_LOCK(:name, 0))');
+        $stmt->bindValue(':name', self::getLockName($key), \PDO::PARAM_STR);
+        $stmt->execute();
+        $result = $stmt->fetchColumn();
+
+        // lock acquired
+        if (1 === $result) {
+            self::$locksAcquired[$id] = true;
+
+            return;
+        }
+
+        if (0 === $result) {
+            throw new LockConflictedException('Lock already acquired by other connection.');
+        }
+
+        if (-1 === $result) {
+            throw new LockConflictedException('Lock already acquired by this connection.');
+        }
+
+        throw new LockAcquiringException('Failed to acquire lock due to mysql error.');
+    }
+
+    public function putOffExpiration(Key $key, float $ttl): void
+    {
+        // noop - GET_LOCK() does not have a ttl
+    }
+
+    public function delete(Key $key): void
+    {
+        $stmt = $this->conn->prepare('DO RELEASE_LOCK(:name)');
+        $stmt->bindValue(':name', self::getLockName($key), \PDO::PARAM_STR);
+        $stmt->execute();
+
+        unset(self::$locksAcquired[$this->getLockId($key)]);
+    }
+
+    public function exists(Key $key): bool
+    {
+        return self::$locksAcquired[$this->getLockId($key)] ?? false;
```

Comment:
That's wrong, the implementation must assert that the lock is still present in the backend. (Ie the database restart, the network is down, ...)

Reply:
This wont be possible since GET_LOCK is bound to the connection so its always released when the connection is terminated.

Reply:
Well.. `getLockId` is based on `connection_id()` which is (for MySQL 8 at lease) an auto-increment that is reset when server starts. 

This means, that each time the server restarts, the application might retrieve the `connection_id` previously affected to another process. Which makes the key used in `self::$locksAcquired` untrustable.

Reply:
I think this should do it. 

I cant add this test case to the since it requires rebooting mysql but I did run it manually to confirm proper handling even if the connection id is the same

```
class ManualMysqlStore extends MysqlStoreTest
{
    public function testMysqlClosedConnection()
    {
        // manually restart mysql to reset conn id counter
        $pdo = $this->getPdo();
        $storeA = new MysqlStore($pdo);
        $key = new Key('foo');
        $storeA->save($key);

        $connIdA = $pdo->query('SELECT CONNECTION_ID()')->fetchColumn();

        while (true) {
            try {
                $this->assertTrue($storeA->exists($key));
                sleep(1);
                // manually restart mysql to reset conn id counter
            } catch (\PDOException $e) {
                // mysql has gone away
                break;
            }
        }

        $pdo = $this->getPdo();
        $storeB = new MysqlStore($pdo);
        $connIdB = $pdo->query('SELECT CONNECTION_ID()')->fetchColumn();

        $this->assertFalse($storeB->exists($key));
        $this->assertSame($connIdA, $connIdB);
    }
}